index.ts

```diff
@@ -75,3 +75,50 @@ export default function diff(
 	}
 	return diffs;
 }
+
+export function patch(
+    obj: Record<string, any> | any[],
+    diffs: Difference[]
+): Record<string, any> | any[] {
+    var newObj = JSON.parse(JSON.stringify(obj)); // deep clone the source object
+    var arrayDelQueue = []
+    var removeSymbol = Symbol()
+
+    for (const diff of diffs) {
+        if (!diff.path || diff.path.length === 0)
+            continue;
+
+        var currObj = newObj;
+        var diffPathLength = diff.path.length;
+        var lastPathElement = diff.path[diffPathLength - 1];
+        var secondLastPathElement = diff.path[diffPathLength - 2]
+        for (var i = 0; i < diffPathLength - 1; i++) {
+            currObj = currObj[diff.path[i]];
+        }
+
+        switch(diff.type) {
+            case "CREATE": // fall-through - equal to: case "CREATE" || "CHANGE"
+            case "CHANGE":
+                currObj[lastPathElement] = diff.value;
+                break;
+            case "REMOVE":
+                if (Array.isArray(currObj)) {
+                    (currObj as any)[lastPathElement] = removeSymbol;
+                    arrayDelQueue.push(() => {
+                        if (secondLastPathElement !== undefined) {
+                            (currObj as any)[secondLastPathElement] = (currObj as any)[secondLastPathElement].filter((e: any) => e !== removeSymbol);
```

Comment:
There are two errors here.

1) you `secondLastPathElement` should be indexed on `currObj`'s parent, not on `currObj`.  The easiest solution is to also keep track of `parent` in the for loop on line 96.

2) the values of `parent` and `secondLastPathElement` needs to be captured in a closure in order to avoid getting overwritten if there are multiple removals in the loop.

Here's my current version, which also checks to ensure that the element matches `oldValue` before removing it.

```typescript
import diff from "microdiff"

export interface Difference {
	type: "CREATE" | "REMOVE" | "CHANGE";
	path: (string | number)[];
	value?: any;
	oldValue?: any;
}

function deepEquals(lhs: any, rhs: any): boolean {
    return lhs === rhs || (lhs !== undefined && rhs !== undefined && diff(lhs, rhs).length == 0)
}

export default function applyDiff(obj: Record<string, any>, diffs: Difference[]): Record<string, any> {
    var newObj = JSON.parse(JSON.stringify(obj)); // deep clone the source object
    var arrayDelQueue = []
    var removeSymbol = Symbol()

    for (const diff of diffs) {
        if (!diff.path || diff.path.length === 0)
            continue;

        var currObj = newObj;
        var parent: any;

        var diffPathLength = diff.path.length;
        var lastPathElement = diff.path[diffPathLength - 1];
        var secondLastPathElement = diff.path[diffPathLength - 2]

        for (var i = 0; i < diffPathLength - 1; i++) {
            parent = currObj;
            currObj = currObj[diff.path[i]];
        }

        switch(diff.type) {
            case "CREATE": // fall-through - equal to: case "CREATE" || "CHANGE"
            case "CHANGE":
                currObj[lastPathElement] = diff.value;
                break;
            case "REMOVE":
                if (Array.isArray(currObj)) {
                    if (deepEquals((currObj as any)[lastPathElement], diff.oldValue)) {
                        (currObj as any)[lastPathElement] = removeSymbol;
                        console.log("Pushing into queue currObj", currObj)
                        const deleteFunc = (parent: any, secondLastPathElement: any) => {
                            return () => {
                                if (secondLastPathElement !== undefined) {
                                    console.log("HERE secondLastPathElement=", secondLastPathElement, ", filtering ",  (parent as any));
                                    (parent as any)[secondLastPathElement] = (parent as any)[secondLastPathElement].filter((e: any) => e !== removeSymbol);
                                } else {
                                    newObj = newObj.filter((e: any) => e !== removeSymbol);
                                }
                            }
                        }
                        arrayDelQueue.push(deleteFunc(parent, secondLastPathElement))
                    }
                } else {
                    if (deepEquals(currObj[lastPathElement], diff.oldValue)) {
                        delete currObj[lastPathElement];
                    }
                }
                break;
        }
    }

    arrayDelQueue.forEach(arrayDeletion => arrayDeletion())

    return newObj;
}

```